lib/federation.js

```diff
@@ -173,6 +175,37 @@ function addTypeNameToResult (result, typename) {
   return result
 }
 
+function resolveType (reference, context, info) {
+  const { __typename } = reference
+
+  const type = info.schema.getType(__typename)
+
+  if (!type || !(isObjectType(type) || isInterfaceType(type))) {
```

Comment:
The first check verifies that the type is allowed to continue the type resolution (should be an object or an interface). The second if discriminate between the object and the interface, when is an object type there is no need to call `resolveType` function because the typename is already defined.